py/server/deephaven/table.py

```diff
@@ -62,6 +69,102 @@ def _j_py_script_session() -> _JPythonScriptSession:
         return None
 
 
+_numpy_type_codes = ["i", "l", "h", "f", "d", "b", "?", "O"]
+
+
+def _encode_signature(fn: Callable) -> str:
+    """Encode the signature of a Python function by mapping the annotations of the parameter types and the return
+    type to numpy dtype chars (i,l,h,f,d,b,?,O), and pack them into a string with parameter type chars first,
+    in their original order, followed by the delimiter string '->', then the return type_char.
+
+    If a parameter or the return of the function is not annotated, the default 'O' - object type, will be used.
+    """
+    sig = inspect.signature(fn)
+
+    parameter_types = []
+    for n, p in sig.parameters.items():
+        try:
+            np_dtype = np.dtype(p.annotation if p.annotation else "object")
+            parameter_types.append(np_dtype)
+        except TypeError:
+            parameter_types.append(np.dtype("object"))
+
+    try:
+        return_type = np.dtype(sig.return_annotation if sig.return_annotation else "object")
+    except TypeError:
+        return_type = np.dtype("object")
+
+    np_type_codes = [np.dtype(p).char for p in parameter_types]
+    np_type_codes = [c if c in _numpy_type_codes else "O" for c in np_type_codes]
+    return_type_code = np.dtype(return_type).char
+    return_type_code = return_type_code if return_type_code in _numpy_type_codes else "O"
+
+    np_type_codes.extend(["-", ">", return_type_code])
+    return "".join(np_type_codes)
+
+
+class DhVectorize:
+    """A decorator to vectorize a Python function used in Deephaven query formulas and invoked on a row basis.
+
+    The result callable will accept arrays of the declared parameter data types (vs. scalar) and iterate through them
+    together and call the wrapped function on each iteration and pack the result of each call into an array and return
+    the result array upon exhausting the array parameters.
```

Comment:
Could probably be worded more clearly.  Also, it doesn't say if the vectorized function can still be used on scalar data.